File: backend/autopilot.py
import asyncio
import time
from predict import get_ai_prediction
from main import state

async def run_autopilot(user_goal):
    page = state["page"]
    task_cat = state["current_task"]
    
    for step in range(15): # Max steps
        # 1. Capture current visual
        shot_path = f"screenshots/auto_{int(time.time())}.png"
        await page.screenshot(path=shot_path)
        
        # 2. AI decides next step based on screenshots and goal
        code = await get_ai_prediction(shot_path, user_goal, task_cat)
        
        if "DONE" in code.upper():
            break
            
        # 3. Execute Keyboard or Mouse commands
        try:
            exec_globals = {"page": page, "asyncio": asyncio}
            clean_code = code.replace("```python", "").replace("```", "").strip()
            exec(f"async def action():\n    {clean_code}", exec_globals)
            await exec_globals["action"]()
            await asyncio.sleep(2) # Let page settle
        except Exception as e:
            print(f"Error: {e}")
            break

File: backend/autopilot.js
import { getAiPrediction } from './predict.js';
import fs from 'fs';
import path from 'path';

export async function runAutopilot(page, userGoal, taskCategory) {
    console.log(`🚀 Starting Auto-Pilot: ${taskCategory}`);
    
    for (let step = 0; step < 15; step++) {
        const timestamp = Date.now();
        const shotPath = path.join('screenshots', `auto_${timestamp}.png`);
        
        // 1. Capture current visual
        await page.screenshot({ path: shotPath });
        
        // 2. Get AI suggestion
        const suggestion = await getAiPrediction(shotPath, userGoal, taskCategory);
        console.log(`🤖 AI Suggestion: ${suggestion}`);

        if (suggestion.includes('TASK_COMPLETE')) break;

        // 3. Execute the code string
        try {
            const cleanCode = suggestion.replace(/```javascript|```js|```/g, '').trim();
            // Using an AsyncFunction constructor to execute the string safely
            const AsyncFunction = Object.getPrototypeOf(async function(){}).constructor;
            const executeAction = new AsyncFunction('page', cleanCode);
            await executeAction(page);
            
            // Wait for page to settle
            await new Promise(r => setTimeout(r, 2000));
        } catch (err) {
            console.error("❌ Execution Error:", err);
            break;
        }
    }
}